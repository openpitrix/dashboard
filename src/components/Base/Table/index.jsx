import React from 'react';
import RcTable from 'rc-table';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import _, { isEqual, find } from 'lodash';
import { translate } from 'react-i18next';
import { withRouter } from 'react-router-dom';
import qs from 'query-string';
import { qs2Obj } from 'utils';

import {
  Checkbox, Radio, Popover, Icon, Pagination
} from 'components/Base';
import Loading from 'components/Loading';
import NoData from './noData';

import styles from './index.scss';

@translate()
export class Table extends React.Component {
  static propTypes = {
    columns: PropTypes.array,
    dataSource: PropTypes.array,
    filterList: PropTypes.array,
    isLoading: PropTypes.bool,
    pagination: PropTypes.object,
    prefixCls: PropTypes.string,
    rowSelection: PropTypes.shape({
      selectedRowKeys: PropTypes.array,
      onChange: PropTypes.func,
      onSelect: PropTypes.func,
      onSelectAll: PropTypes.func
    })
  };

  static defaultProps = {
    prefixCls: 'pi-table',
    dataSource: [],
    columns: [],
    pagination: {},
    rowKey: 'key',
    rowSelection: {
      // selectedRowKeys: [],
      onChange: _.noop,
      onSelect: _.noop,
      onSelectAll: _.noop
    },
    filterList: [],
    isLoading: false
  };

  constructor(props) {
    super(props);

    const { location } = props;
    const values = qs2Obj(location.search);
    const reverse = values.reverse !== '0';
    this.state = {
      selectedRowKeys: (props.rowSelection || {}).selectedRowKeys || [],
      selectionDirty: false,
      sort_key: values.sort_key,
      reverse
    };
  }

  componentWillReceiveProps(nextProps) {
    if (!isEqual(this.props.dataSource, nextProps.dataSource)) {
      this.setState({
        selectionDirty: false
      });
    }

    if (nextProps.rowSelection && nextProps.rowSelection.selectedRowKeys) {
      this.setState({
        selectedRowKeys: nextProps.rowSelection.selectedRowKeys || []
      });
    }
  }

  getTableData = () => {
    const { rowKey } = this.props;
    const { reverse, sort_key } = this.state;
    const order = reverse ? 'desc' : 'asc';
    const data = _.orderBy(this.props.dataSource, sort_key, order);

    return data.map((item, i) => {
      if (!item[rowKey]) {
        item[rowKey] = item.id || rowKey + i;
      }
      return item;
    });
  };

  getItemKey = (item, index) => {
    const rowKey = this.props.rowKey;
    const itemKey = typeof rowKey === 'function' ? rowKey(item, index) : item[rowKey];
    return itemKey === undefined ? index : itemKey;
  };

  getDefaultSelection = () => this.getTableData()
    .filter(item => item.checked)
    .map((item, i) => this.getItemKey(item, i));

  setSelectedRowKeys = (
    selectedRowKeys,
    {
      selectType, item, checked, changeRowKeys
    }
  ) => {
    const { rowSelection } = this.props;
    this.setState({
      selectedRowKeys,
      selectionDirty: true
    });

    const data = this.getTableData();

    if (!rowSelection.onChange && !rowSelection[selectType]) {
      return;
    }
    const selectedRows = data.filter(
      (row, i) => selectedRowKeys.indexOf(this.getItemKey(row, i)) !== -1
    );
    if (rowSelection.onChange) {
      rowSelection.onChange(selectedRowKeys, selectedRows);
    }

    if (rowSelection[selectType]) {
      switch (selectType) {
        default:
        case 'onSelect':
          rowSelection[selectType](checked, item, selectedRows);
          break;
        case 'onSelectAll': {
          const changeRows = data.filter(
            (row, i) => changeRowKeys.indexOf(this.getItemKey(row, i)) !== -1
          );
          rowSelection[selectType](checked, selectedRows, changeRows);
          break;
        }
        case 'onSelectInvert':
          rowSelection[selectType](selectedRowKeys);
          break;
      }
    }
  };

  handleCheckboxSelect = (value, index, e) => {
    const { rowKey } = this.props;
    const checked = e.target.checked;
    const defaultSelection = this.state.selectionDirty
      ? []
      : this.getDefaultSelection();
    let selectedRowKeys = this.state.selectedRowKeys.concat(defaultSelection);

    if (checked) {
      selectedRowKeys.push(value[rowKey]);
    } else {
      selectedRowKeys = selectedRowKeys.filter(key => value[rowKey] !== key);
    }

    this.setSelectedRowKeys(selectedRowKeys, {
      selectType: 'onSelect',
      item: value,
      checked
    });
  };

  handleRadioSelect = () => {};

  handleSelectAll = e => {
    const type = e.target.checked ? 'all' : 'removeAll';
    const data = this.getTableData();
    const defaultSelection = this.state.selectionDirty
      ? []
      : this.getDefaultSelection();
    const selectedRowKeys = this.state.selectedRowKeys.concat(defaultSelection);
    const changeableRowKeys = data
      .filter(item => !item.disabled)
      .map((item, i) => this.getItemKey(item, i));

    const changeRowKeys = [];
    let selectType = '';
    let checked;

    switch (type) {
      default:
      case 'all':
        changeableRowKeys.forEach(key => {
          if (selectedRowKeys.indexOf(key) === -1) {
            selectedRowKeys.push(key);
            changeRowKeys.push(key);
          }
        });
        selectType = 'onSelectAll';
        checked = true;
        break;
      case 'removeAll':
        changeableRowKeys.forEach(key => {
          if (selectedRowKeys.indexOf(key) !== -1) {
            selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1);
            changeRowKeys.push(key);
          }
        });
        selectType = 'onSelectAll';
        checked = false;
        break;
      case 'invert':
        changeableRowKeys.forEach(key => {
          if (selectedRowKeys.indexOf(key) === -1) {
            selectedRowKeys.push(key);
          } else {
            selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1);
          }
          changeRowKeys.push(key);
        });
        selectType = 'onSelectInvert';
        break;
    }

    this.setSelectedRowKeys(selectedRowKeys, {
      selectType,
      checked,
      changeRowKeys
    });
  };

  renderSelectionBox = type => (value, row, index) => {
    const handleChange = e => {
      e.preventDefault();
      type === 'radio'
        ? this.handleRadioSelect(value, index, e)
        : this.handleCheckboxSelect(value, index, e);
    };

    if (type === 'radio') {
      return <Radio />;
    }

    return (
      <Checkbox
        disabled={value.disabled}
        checked={
          this.state.selectedRowKeys.indexOf(value[this.props.rowKey]) !== -1
        }
        onChange={handleChange}
      />
    );
  };

  renderHeaderColumns = () => {
    const { rowSelection, columns } = this.props;
    const tableColumns = [...columns];

    if (rowSelection.selectedRowKeys) {
      const data = this.getTableData();
      const selectionColumn = {
        key: 'selection-column',
        width: '30px',
        render: this.renderSelectionBox(rowSelection.type),
        className: styles.selectionColumn,
        fixed: rowSelection.fixed
      };

      if (rowSelection.type === 'radio') {
        selectionColumn.title = '';
      } else {
        const allChecked = data.length
          ? data.every(
            (item, i) => this.state.selectedRowKeys.indexOf(this.getItemKey(item, i))
                !== -1
          )
          : false;
        selectionColumn.title = (
          <Checkbox checked={allChecked} onChange={this.handleSelectAll} />
        );
      }

      if (tableColumns[0] && tableColumns[0].key === 'selection-column') {
        tableColumns[0] = selectionColumn;
      } else {
        tableColumns.unshift(selectionColumn);
      }
    }
    return tableColumns;
  };

  renderFilterContent = filter => {
    const { t } = this.props;

    return (
      <ul className="filterContent">
        {filter.conditions.map(condition => (
          <li
            key={condition.value}
            onClick={() => filter.onChangeFilter(condition.value)}
            className={classNames({
              active: condition.value === filter.selectValue
            })}
          >
            {t(condition.name)}
          </li>
        ))}
      </ul>
    );
  };

  trans = key => {
    if (typeof key === 'string') {
      return this.props.t(key);
    }
    return key;
  };

  handleSort = (cb, column) => {
    const sort_key = column.key || column.dataIndex;
    this.setState(
      {
        sort_key,
        reverse: !this.state.reverse
      },
      () => {
        const { reverse } = this.state;
        const { history, location } = this.props;
        const values = qs2Obj(location.search);
        values.reverse = reverse ? 1 : 0;
        values.sort_key = sort_key;
        history.push({
          search: qs.stringify(values)
        });

        typeof cb === 'function'
          && cb({
            sort_key: column.key || column.dataIndex,
            reverse
          });
      }
    );
  };

  filterColumns = columns => {
    const { filterList, t } = this.props;
    const { reverse } = this.state;

    return columns.map(({ ...column }) => {
      const filter = find(filterList, { key: column.key });

      if (filter) {
        column.title = (
          <Popover
            content={this.renderFilterContent(filter)}
            className={styles.filterOuter}
          >
            {this.trans(column.title)}
            <Icon
              name="caret-down"
              type={`${filter.selectValue ? 'light' : 'dark'}`}
              size={12}
            />
          </Popover>
        );
      } else if (typeof column.title === 'string') {
        // won't call translate helper in columns file
        column.title = t(column.title);
      }

      // is col can be sorted?
      if (column.sorter) {
        column.title = (
          <span
            onClick={() => this.handleSort(column.onChangeSort, column)}
            className={styles.sortOuter}
          >
            {column.title}
            <Icon name={`sort-${reverse ? 'descend' : 'ascend'}ing`} />
          </span>
        );
      }

      return column;
    });
  };

  rowClassName = (record, index) => {
    let { selectedRowKeys } = this.state;
    if ('toJSON' in selectedRowKeys) {
      selectedRowKeys = selectedRowKeys.toJSON();
    }
    if (selectedRowKeys.includes(index)) {
      return 'row-selected';
    }
    return '';
  };

  renderTable = () => {
    const { pagination, ...restProps } = this.props;
    const data = this.getTableData();
    let columns = this.renderHeaderColumns();
    columns = this.filterColumns(columns);

    return (
      <RcTable
        {...restProps}
        key="table"
        data={data}
        columns={columns}
        emptyText={<NoData type={pagination.tableType} />}
        rowClassName={this.rowClassName}
      />
    );
  };

  render() {
    const {
      className,
      style,
      pagination,
      isLoading,
      rowSelection
    } = this.props;
    const { selectedRowKeys } = this.state;

    return (
      <div className={classNames(styles.table, className)} style={style}>
        <Loading isLoading={isLoading}>
          {this.renderTable()}
          <Pagination
            {...pagination}
            selectedRows={selectedRowKeys}
            onSelectRow={rowSelection.onChange}
          />
        </Loading>
      </div>
    );
  }
}

export default withRouter(Table);
