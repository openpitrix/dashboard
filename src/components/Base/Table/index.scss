@import '~scss/vars';

.table {
  &-body {
    transition: opacity $transition-speed;
  }

  table {
    width: 100%;
    border-collapse: collapse;
    border-spacing: 0;
    text-align: left;
    box-sizing: border-box;
    overflow: hidden;

    .selectionColumn {
      width: 14px;
      padding: 0 7px;
      label {
        display: none;
        padding: 0;
      }
      [class*='checked'] {
        display: inline-block;
      }
    }

    tr:hover {
      .selectionColumn {
        label {
          display: inline-block;
        }
      }
    }

    th {
      .filterOuter {
        max-width: 120px;
        [class*='target'] {
          height: auto;
          //width: 60px;
          width: 68px;
          border: 0 none;
          box-shadow: none;
          &:hover {
            border: 0 none;
            box-shadow: none;
          }
        }
        :global {
          .pi-popover-popper {
            //margin-left: -52px;
            margin-left: -42px;
          }
          .filterContent {
            li {
              padding: 0 12px;
              width: 80px;
              line-height: 32px;
              font-size: 14px;
              font-weight: normal;
              color: $N300;
              cursor: pointer;
              transition: all $transition-speed;
              &:hover,
              &.active {
                background-color: $N10;
                color: $N500;
              }
            }
          }
          .icon {
            float: right;
            /*position: relative;
            top: 3px;
            margin-left: 8px;*/
          }
        }
      }
      .sortOuter {
        :global {
          .icon {
            vertical-align: middle;
            margin-left: 6px;
            cursor: pointer;
          }
        }
      }
    }
  }
}

:global {
  .pi-table {
    overflow-x: auto;

    &-thead > tr > th {
      position: relative;
      padding: 10px 5px;
      line-height: 12px;
      font-weight: normal;
      font-size: 12px;
      color: $N100;
      box-shadow: inset 0 -1px 0 0 $N10, inset 0 1px 0 0 $N10;
      letter-spacing: 0;

      &.actions {
        text-align: center;
      }

      &::before {
        content: '';
        position: absolute;
        left: -1px;
        top: 8px;
        bottom: 8px;
        width: 0;
      }

      label {
        margin-left: 0 !important;
      }

      //.pi-popover-popper[data-pacement^="bottom"] {
      //  transform: translate3d(500px, 236px, 0px) !important;
      //}
    }

    &-tbody > tr {
      transition: all $transition-speed;
      &:hover {
        background-color: $tr-hover-bg;
        box-shadow: 0 1px 4px 0 rgba(73, 33, 173, 0.06),
          0 4px 8px 0 rgba(35, 35, 36, 0.04);

        td {
          &:first-child {
            border-left-color: $tr-hover-bg;
          }
          &:last-child {
            border-right-color: $tr-hover-bg;
          }
        }

        td a,
        td [class*='tdName'] a,
        td > a > [class*='repoName'] {
          color: $link-hover;
        }

        td [class*='popper'] a {
          color: $N500;
        }

        td.actions > div{
         display: block;
        }
      }

      &:last-child {
        td {
          //border-bottom: none;
        }
      }
    }

    &-tbody > tr > td {
      padding: 14px 5px;
      line-height: 16px;
      box-shadow: inset 0 -1px 0 0 $N10;
      transition: all $transition-speed;
      word-wrap: break-word;
      word-break: break-all;
      font-size: 12px;

      a {
        color: $N500;
      }
      &:first-child {
        //padding-left: 0;
        //border-left: 48px solid #fff;
      }

      &:last-child {
        //border-right: 48px solid #fff;
      }

      &.actions > div {
        display: none;
      }
    }

    &-placeholder {
      padding: 12px 0;
      text-align: center;
    }
  }
}

.noData {
  padding: 20px 0;
  text-align: center;
  .word {
    margin-bottom: 16px;
    font-size: 12px;
    line-height: 1.67;
    letter-spacing: 0;
    color: $N100;
  }
  :global {
    .icon {
      opacity: 0.2;
    }
  }
}
