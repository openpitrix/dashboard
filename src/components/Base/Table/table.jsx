import React from 'react';
import RcTable from 'rc-table';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { isEqual } from 'lodash';

import {
  Checkbox, Radio, Popover, Icon, Pagination
} from 'components/Base';
import Loading from 'components/Loading';
import NoData from './noData';

import styles from './index.scss';

export default class Table extends React.Component {
  static propTypes = {
    columns: PropTypes.array,
    dataSource: PropTypes.array,
    filterList: PropTypes.array,
    isLoading: PropTypes.bool,
    pagination: PropTypes.object,
    prefixCls: PropTypes.string,
    rowSelection: PropTypes.object
  };

  static defaultProps = {
    prefixCls: 'pi-table',
    dataSource: [],
    columns: [],
    pagination: {},
    rowKey: 'key',
    rowSelection: {}
  };

  constructor(props) {
    super(props);

    this.state = {
      selectedRowKeys: (props.rowSelection || {}).selectedRowKeys || [],
      selectionDirty: false,
      reverse: true
    };
  }

  getTableData = () => {
    const { rowKey } = this.props;
    const data = [...this.props.dataSource];

    data.map((item, i) => {
      if (!item[rowKey]) {
        item[rowKey] = item.id || rowKey + i;
      }
      return item;
    });
    return data;
  };

  getItemKey = (item, index) => {
    const rowKey = this.props.rowKey;
    const itemKey = typeof rowKey === 'function' ? rowKey(item, index) : item[rowKey];
    return itemKey === undefined ? index : itemKey;
  };

  getDefaultSelection = () => this.getTableData()
    .filter(item => item.checked)
    .map((item, i) => this.getItemKey(item, i));

  setSelectedRowKeys = (
    selectedRowKeys,
    {
      selectType, item, checked, changeRowKeys
    }
  ) => {
    const { rowSelection } = this.props;
    this.setState({
      selectedRowKeys,
      selectionDirty: true
    });

    const data = this.getTableData();

    if (!rowSelection.onChange && !rowSelection[selectType]) {
      return;
    }
    const selectedRows = data.filter(
      (row, i) => selectedRowKeys.indexOf(this.getItemKey(row, i)) !== -1
    );
    if (rowSelection.onChange) {
      rowSelection.onChange(selectedRowKeys, selectedRows);
    }

    if (rowSelection[selectType]) {
      switch (selectType) {
        default:
        case 'onSelect':
          rowSelection[selectType](checked, item, selectedRows);
          break;
        case 'onSelectAll': {
          const changeRows = data.filter(
            (row, i) => changeRowKeys.indexOf(this.getItemKey(row, i)) !== -1
          );
          rowSelection[selectType](checked, selectedRows, changeRows);
          break;
        }
        case 'onSelectInvert':
          rowSelection[selectType](selectedRowKeys);
          break;
      }
    }
  };

  setExtendRowKeys = (
    selectedRowKeys,
    {
      selectType, item, checked, changeRowKeys
    }
  ) => {
    const { extendRowSelection } = this.props;
  };

  handleCheckboxSelect = (value, index, e) => {
    const { rowKey } = this.props;
    const checked = e.target.checked;
    const defaultSelection = this.state.selectionDirty
      ? []
      : this.getDefaultSelection();
    let selectedRowKeys = this.state.selectedRowKeys.concat(defaultSelection);

    if (checked) {
      selectedRowKeys.push(value[rowKey]);
    } else {
      selectedRowKeys = selectedRowKeys.filter(key => value[rowKey] !== key);
    }

    this.setSelectedRowKeys(selectedRowKeys, {
      selectType: 'onSelect',
      item: value,
      checked
    });
  };

  handleRadioSelect = () => {};

  handleSelectAll = e => {
    const type = e.target.checked ? 'all' : 'removeAll';
    const data = this.getTableData();
    const defaultSelection = this.state.selectionDirty
      ? []
      : this.getDefaultSelection();
    const selectedRowKeys = this.state.selectedRowKeys.concat(defaultSelection);
    const changeableRowKeys = data
      .filter(item => !item.disabled)
      .map((item, i) => this.getItemKey(item, i));

    const changeRowKeys = [];
    let selectType = '';
    let checked;

    switch (type) {
      default:
      case 'all':
        changeableRowKeys.forEach(key => {
          if (selectedRowKeys.indexOf(key) === -1) {
            selectedRowKeys.push(key);
            changeRowKeys.push(key);
          }
        });
        selectType = 'onSelectAll';
        checked = true;
        break;
      case 'removeAll':
        changeableRowKeys.forEach(key => {
          if (selectedRowKeys.indexOf(key) !== -1) {
            selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1);
            changeRowKeys.push(key);
          }
        });
        selectType = 'onSelectAll';
        checked = false;
        break;
      case 'invert':
        changeableRowKeys.forEach(key => {
          if (selectedRowKeys.indexOf(key) === -1) {
            selectedRowKeys.push(key);
          } else {
            selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1);
          }
          changeRowKeys.push(key);
        });
        selectType = 'onSelectInvert';
        break;
    }

    this.setSelectedRowKeys(selectedRowKeys, {
      selectType,
      checked,
      changeRowKeys
    });
  };

  renderSelectionBox = type => (value, row, index) => {
    const handleChange = e => {
      e.stopPropagation();
      type === 'radio'
        ? this.handleRadioSelect(value, index, e)
        : this.handleCheckboxSelect(value, index, e);
    };

    if (type === 'radio') {
      return <Radio />;
    }

    return (
      <Checkbox
        disabled={value.disabled}
        checked={
          this.state.selectedRowKeys.indexOf(value[this.props.rowKey]) !== -1
        }
        onChange={handleChange}
      />
    );
  };

  renderRowSelection = () => {
    const { rowSelection, columns } = this.props;
    const tableColumns = [...columns];

    if (rowSelection.selectedRowKeys) {
      const data = this.getTableData();
      const selectionColumn = {
        key: 'selection-column',
        width: '30px',
        render: this.renderSelectionBox(rowSelection.type),
        className: styles.selectionColumn,
        fixed: rowSelection.fixed
      };

      if (rowSelection.type === 'radio') {
        selectionColumn.title = '';
      } else {
        const allChecked = data.length
          ? data.every(
            (item, i) => this.state.selectedRowKeys.indexOf(this.getItemKey(item, i))
                !== -1
          )
          : false;
        selectionColumn.title = (
          <Checkbox checked={allChecked} onChange={this.handleSelectAll} />
        );
      }

      if (tableColumns[0] && tableColumns[0].key === 'selection-column') {
        tableColumns[0] = selectionColumn;
      } else {
        tableColumns.unshift(selectionColumn);
      }
    }
    return tableColumns;
  };

  renderFilterContent = filter => (
    <ul className="filterContent">
      {filter.conditions.map(condition => (
        <li
          key={condition.value}
          onClick={() => filter.onChangeFilter(condition.value)}
          className={classNames({
            active: condition.value === filter.selectValue
          })}
        >
          {condition.name}
        </li>
      ))}
    </ul>
  );

  renderFilterColumn = columns => {
    const { filterList } = this.props;
    if (filterList) {
      columns = columns.map(column => {
        const newColumn = { ...column };
        const filter = filterList.find(
          element => element.key === newColumn.key
        );
        if (filter) {
          newColumn.title = (
            <Popover
              content={this.renderFilterContent(filter)}
              className={styles.filterOuter}
            >
              {newColumn.title}
              <Icon
                name="caret-down"
                type={`${filter.selectValue ? 'light' : 'dark'}`}
              />
            </Popover>
          );
        }
        return newColumn;
      });
    }
    return columns;
  };

  rowClassName = (record, index) => {
    let { selectedRowKeys } = this.state;
    if ('toJSON' in selectedRowKeys) {
      selectedRowKeys = selectedRowKeys.toJSON();
    }
    if (selectedRowKeys.includes(index)) {
      return 'row-selected';
    }
    return '';
  };

  renderTable = () => {
    const { pagination, ...restProps } = this.props;
    const data = this.getTableData();
    let columns = this.renderRowSelection();
    columns = this.renderFilterColumn(columns);
    columns = columns.map((column, i) => {
      const newColumn = { ...column };
      newColumn.key = column.key || column.dataIndex || i;
      if (newColumn.sorter) {
        const { reverse } = this.state;
        const onChangeSort = () => {
          this.setState({
            reverse: !reverse
          });
          column.onChangeSort({
            sort_key: column.key,
            reverse
          });
        };
        newColumn.title = (
          <span onClick={() => onChangeSort()} className={styles.sortOuter}>
            {newColumn.title}
            <Icon name={`sort-${reverse ? 'ascend' : 'descend'}ing`} />
          </span>
        );
      }
      return newColumn;
    });

    return (
      <RcTable
        {...restProps}
        key="table"
        data={data}
        columns={columns}
        emptyText={<NoData type={pagination.tableType} />}
        rowClassName={this.rowClassName}
      />
    );
  };

  componentWillReceiveProps(nextProps) {
    if (!isEqual(this.props.dataSource, nextProps.dataSource)) {
      this.setState({
        selectionDirty: false
      });
    }

    if (nextProps.rowSelection && nextProps.rowSelection.selectedRowKeys) {
      this.setState({
        selectedRowKeys: nextProps.rowSelection.selectedRowKeys || []
      });
    }
  }

  render() {
    const {
      className,
      style,
      pagination,
      isLoading,
      rowSelection
    } = this.props;
    const { selectedRowKeys } = this.state;

    if (isLoading) {
      return <Loading className="loadTable" isLoading={isLoading} />;
    }

    return (
      <Loading className="loadTable" isLoading={isLoading}>
        <div className={classNames(styles.table, className)} style={style}>
          {this.renderTable()}
          <Pagination
            {...pagination}
            selectedRows={selectedRowKeys}
            onSelectRow={rowSelection.onChange}
          />
        </div>
      </Loading>
    );
  }
}
