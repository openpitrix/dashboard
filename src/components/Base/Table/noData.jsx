import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { translate } from 'react-i18next';

import { Icon } from 'components/Base';
import styles from './index.scss';

@translate()
export default class NoData extends PureComponent {
  static propTypes = {
    type: PropTypes.string
  };

  static defaultProps = {
    type: 'Apps'
  };

  render() {
    const { type, t } = this.props;
    const iconMap = {
      Apps: 'appcenter',
      Clusters: 'cluster',
      Runtimes: 'stateful-set',
      Repos: 'stateful-set',
      Users: 'group',
      Categories: 'components',
      Events: 'tateful-set',
      Versions: 'tag'
    };
    const iconName = iconMap[type] || 'appcenter';

    return (
      <div className={styles.noData}>
        <Icon name={iconName} size={120} type={`light`} />
        <div className={styles.word}>{t('No search data')}</div>
      </div>
    );
  }
}
