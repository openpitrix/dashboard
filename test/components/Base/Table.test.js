import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Table from 'components/Base/Table';

const setup = (type, props = {}) => (type === 'render'
  ? render(<Table {...props} />)
  : mount(<Table {...props} />));

describe('Base/Table', () => {
  const columns = [
    {
      title: 'Name',
      dataIndex: 'name',
      key: 'name',
      width: '50%'
    },
    {
      title: 'Age',
      dataIndex: 'age',
      key: 'age',
      width: '50%'
    }
  ];
  const dataSource = [
    {
      name: 'zhangsan',
      age: 20,
      key: '1'
    },
    {
      name: 'lisi',
      age: 22,
      key: '2'
    }
  ];
  const rowSelection = {
    selectedRowKeys: [],
    onChange: jest.fn(),
    onSelect: jest.fn(),
    onSelectAll: jest.fn()
  };
  const pagination = {
    tableType: 'Clusters',
    onChange: jest.fn(),
    total: 2,
    current: 2
  };
  it('basic render', () => {
    const wrapper = setup('render', { columns, dataSource });

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('call onChange', () => {
    const wrapper = setup('mount', {
      columns,
      dataSource,
      rowSelection,
      pagination
    });
    const checkbox = wrapper
      .find('tbody tr')
      .first()
      .find('td')
      .at(0)
      .find('input[type="checkbox"]');
    checkbox.simulate('change', { target: { checked: true } });

    const mockChange = rowSelection.onChange;
    expect(wrapper.state().selectionDirty).toBeTruthy();
    expect(mockChange).toHaveBeenCalled();
    expect(mockChange.mock.calls[0][0]).toEqual(['1']);
  });

  it('call onSelect', () => {
    const wrapper = setup('mount', {
      columns,
      dataSource,
      rowSelection,
      pagination
    });
    const checkbox = wrapper
      .find('tbody tr')
      .first()
      .find('td')
      .at(0)
      .find('input[type="checkbox"]');
    checkbox.simulate('change', { target: { checked: true } });

    const mockSelect = rowSelection.onSelect;
    expect(wrapper.state().selectionDirty).toBeTruthy();
    expect(mockSelect).toHaveBeenCalled();
    expect(mockSelect.mock.calls[0][1]).toEqual(dataSource[0]);
  });

  it('call onSelectAll', () => {
    const wrapper = setup('mount', {
      columns,
      dataSource,
      rowSelection,
      pagination
    });
    const checkbox = wrapper
      .find('thead th')
      .at(0)
      .find('input[type="checkbox"]');
    checkbox.simulate('change', { target: { checked: true } });

    const mockSelectAll = rowSelection.onSelectAll;
    expect(wrapper.state().selectionDirty).toBeTruthy();
    expect(mockSelectAll).toHaveBeenCalled();
    expect(mockSelectAll.mock.calls[0][1]).toEqual(dataSource);
  });
});
