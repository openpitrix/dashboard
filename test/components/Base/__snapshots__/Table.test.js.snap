// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Table basic render 1`] = `
<div
  class="table"
>
  <div
    class="pi-table pi-table-scroll-position-left"
  >
    <div
      class="pi-table-content"
    >
      <div
        class="pi-table-body"
      >
        <table
          class=""
        >
          <colgroup>
            <col
              style="width:50%;min-width:50%"
            />
            <col
              style="width:50%;min-width:50%"
            />
          </colgroup>
          <thead
            class="pi-table-thead"
          >
            <tr>
              <th
                class=""
              >
                Name
              </th>
              <th
                class=""
              >
                Age
              </th>
            </tr>
          </thead>
          <tbody
            class="pi-table-tbody"
          >
            <tr
              class="pi-table-row  pi-table-row-level-0"
            >
              <td
                class=""
              >
                <span
                  class="pi-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                zhangsan
              </td>
              <td
                class=""
              >
                20
              </td>
            </tr>
            <tr
              class="pi-table-row  pi-table-row-level-0"
            >
              <td
                class=""
              >
                <span
                  class="pi-table-row-indent indent-level-0"
                  style="padding-left:0px"
                />
                lisi
              </td>
              <td
                class=""
              >
                22
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    class="pagination"
  >
    <div
      class="total"
    >
      : 100
    </div>
    <div
      class="pageItems"
    >
      <span
        aria-disabled="true"
        class="btn btn-disabled"
      >
        <span
          class="icon icon-previous"
          style="width:16px;height:16px"
        >
          <svg
            class="qicon qicon-previous qicon-light"
          >
            <use
              href="#previous"
            />
          </svg>
        </span>
      </span>
      <span
        class="current"
      >
        1 / 10
      </span>
      <span
        aria-disabled="false"
        class="btn"
      >
        <span
          class="icon icon-next"
          style="width:16px;height:16px"
        >
          <svg
            class="qicon qicon-next qicon-light"
          >
            <use
              href="#next"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</div>
`;
